packages/slate-editor/src/extensions/embed/EmbedExtension.tsx

```diff
@@ -17,9 +17,15 @@ interface Parameters extends EmbedExtensionConfiguration {
 
 export const EXTENSION_ID = 'EmbedExtension';
 
+export enum Provider {
+    INSTAGRAM = 'instagram',
+    YOUTUBE = 'youtube',
+}
+
 export interface EmbedExtensionConfiguration {
     fetchOembed: (url: OEmbedInfo['url']) => Promise<OEmbedInfo>;
     showAsScreenshot: boolean;
+    providers: `${Provider}`[];
```

Comment:
Hm... :thinking: 

Now, when I see this, I start to believe this is not the best place for the property. 

These are the smells:

1. `Provider` enum is a full copy of the `PlaceholderNode.Provider`
2. Some of the providers have nothing to do with Embeds and actually produce a `VideoNode` (e.g.: YouTube)
3. The extension itself is also almost completely unrelated to the functionality: no new code is being added to it.

That being said, I believe it would be better to move this setting to the `FloatingAddMenuExtension`, as this setting is mainly controlling the menu options. Right?

Also, let's define the `FloatingAddMenuExtension` config property using the `PlaceholderNode.Provider` enum (to avoid duplication).

What do you think @MohammadxAli ?

Reply:
> 2. Some of the providers have nothing to do with Embeds and actually produce a VideoNode (e.g.: YouTube)

I also felt that it seemed like a code smell when I was implementing it. Looking into the code and I can't find a better place except for the `FloatingAddMenuExtenstion` (as you suggested) to define the `Provider`. But still, mixed feelings.

Let's start with this 👍 I guess as we build we might find a better place for it.

Reply:
> better place except for the FloatingAddMenuExtenstion (as you suggested) to define the Provider.

For the record:

I suggest keeping the `Provider` enum in the `PlaceholderNode` type definition (as is). And just import it in the `FloatingAddMenuExtension`.

Reply:
> better place except for the FloatingAddMenuExtenstion (as you suggested) to define the Provider.

For the record:

I suggest keeping the `Provider` enum in the `PlaceholderNode` type definition (as is). And just import it in the `FloatingAddMenuExtension`.

Reply:
Maybe... Special embeds should be their own extension?

Reply:
What is a special embed?

Reply:
So I have been thinking about this.

Special embeds depend at least on 2 other extensions:
1. `withEmbeds`
2. `withVideos`
3. `withPlaceholders` (optional)

In order to use them, one should have enabled at least `withVideos` and `withEmbeds` extensions to start with because we're supporting a range of possible providers in an array.

So I guess it makes sense to make the special embeds an extension of their own, something like `withSpecialEmbeds` which has 2 dependencies (videos and embeds).

What do you think?

Reply:
> What is a special embed?

"Specific embeds" group as we call them in the UI as well. Instagram, Youtube and etc...

Reply:
They for sure do not need their own extension.

Another thing is that the chosen configuration interface to enable them from outside might be suboptimal. :thinking: 

> withSpecialEmbeds which has 2 dependencies (videos and embeds)

This argument isn't strong. As the menu extension does already depend on almost every other block-type extension. So I wouldn't worry about it.

-- 

How about changing `specialEmbeds?: Provider[]` to `withFloatingAddMenu: { withSpecificProviderOptions?: boolean }`?

and then the menu will generate these additional options depending on the flags:

```ts
if (withEmbeds && withSpecificProviderOptions) {
   yield { /* Instagram menu option */ };
}

if (withVideos && withSpecificProviderOptions) {
   yield { /* YouTube menu option */ };
}
```

I believe this way it should resolve your concern on the extensions inter-dependency. They logic will reflect it with the IFs.